src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,699 @@
+/**
+* Copyright (C) Los Alamos National Security, LLC. 2019 ALL RIGHTS RESERVED.
+* 
+* See file LICENSE for terms.
+*/
+
+#include "ucp_am.h"
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_ep.inl>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/core/ucp_context.h>
+#include <ucp/proto/proto.h>
+#include <ucp/proto/proto_am.inl>
+#include <ucp/dt/dt.h>
+#include <ucp/dt/dt.inl>
+
+void ucp_am_ep_init(ucp_ep_h ep)
+{
+    ucp_ep_ext_proto_t *ep_ext = ucp_ep_ext_proto(ep);
+    
+    if (ep->worker->context->config.features & UCP_FEATURE_EXPERIMENTAL) {
+        ucs_list_head_init(&ep_ext->am.started_ams);
+    }
+}
+
+void ucp_am_ep_cleanup(ucp_ep_h ep)
+{
+    ucp_ep_ext_proto_t *ep_ext = ucp_ep_ext_proto(ep);
+
+    if (ep->worker->context->config.features & UCP_FEATURE_EXPERIMENTAL) {
+        if (ucs_unlikely(!(ucs_list_is_empty(&ep_ext->am.started_ams)))) {
+            ucs_warn("not all UCP active messages have been" 
+                      "run to comepletion");
+        }
+    }
+}
+
+UCS_PROFILE_FUNC_VOID(ucp_am_data_release,
+                      (worker, data),
+                      ucp_worker_h worker, void *data)
+{
+    ucp_recv_desc_t *rdesc = (ucp_recv_desc_t *) data - 1;
+    ucp_recv_desc_t *desc;
+
+    if (rdesc->flags & UCP_RECV_DESC_FLAG_MALLOC) {
+        ucs_free(rdesc);
+        return;
+    } else if (rdesc->flags & UCP_RECV_DESC_FLAG_AM_HDR) {
+        desc = rdesc;
+        rdesc = UCS_PTR_BYTE_OFFSET(rdesc, -sizeof(ucp_am_hdr_t));
+        *rdesc = *desc;
```

Comment:
Forgot how this works, I'll look into it.

Reply:
There is a problem with this, which I will fix, but it still needs to occur. If the descriptor is allocated by UCT, then we need to find out where UCT thinks the descriptor actually starts because of the different header sizes. If UCT did allocate the descriptor, then we overwrite the ucp_am header as the descriptor. In this example, desc points to the part of the ucp_am header that was overwritten, and rdesc points to what ucp_recv_desc_release needs. Before we pass rdesc into ucp_recv_desc_release, we have to rewrite it to be what the descriptor should actually contain, which is desc.

Reply:
Nevermind, there is no problem with this. If you have a better way to do this, I am open to making improvement.

Reply:
"then we overwrite the ucp_am header as the descriptor" - can you point to the code which overwrites it? in general, we should put UCP-specific data in the "rx_headroom" and avoid overwriting UCT descriptor, which is before rx_headroom:
  https://github.com/openucx/ucx/blob/master/src/uct/api/uct_def.h#L57

Reply:
I tried to do that. However the problem arises that the ucp-specific data is different depending on the message type because of the different header sizes. Long does not have this problem because we don't allocate space for the header and if UCT_CB_PARAM_FLAG_DESC is not set, this is not a problem because we tell ucp_recv_desc_init that we do not have a header. If UCT_CB_PARAM_FLAG_DESC is set then UCT has already allocated space for the header so we need to know how much to subtract from the user pointer to get to the start of the UCT descriptor, which is stored as a descriptor flag. So when we call ucp_recv_desc_init we overwrite the ucp_am header with the descriptor, this occurs in ucp_recv_desc_init. We then know that we only have to back up sizeof(ucp_recv_desc_t) and then from there can find how much more to subtract to get to the original descriptor that uct allocated. This was the best way I could think of without changing things elsewhere, again I am open to suggestions though.

Reply:
@snyjm-18 let's add TODO/FIXME comment to try to avoid overriding the recv descriptor (in the place where the actual override happens). I'll take another look after this PR is merged to see if there are better suggestions.

Reply:
I think I have a better way of doing this, but I can do it in another PR.